pkg/controller/controller.go

```diff
@@ -0,0 +1,286 @@
+		remoteServices:      r.remoteServices,
+		federator:           r.federator,
+		remoteServicesMutex: &sync.Mutex{},
+	}
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
+	r.remoteClusters[clusterID] = remoteCluster
+
+	serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+	})
+
+	go remoteCluster.run()
+}
+
+func (r *LightHouseController) removeServiceWatcher(clusterID string) {
+	remoteCluster, ok := r.remoteClusters[clusterID]
+	if ok {
+		klog.V(2).Infof("Stopping watcher for %s", clusterID)
+		remoteCluster.close()
+		delete(r.remoteClusters, clusterID)
+	}
+}
+
+func (r *remoteCluster) close() {
+	close(r.stopCh)
+	r.queue.ShutDown()
+	klog.Infof("Watcher queue shutdown for cluster: %s", r.clusterID)
+}
+
+func (r *remoteCluster) run() {
+	defer utilruntime.HandleCrash()
+	go r.serviceInformer.Run(r.stopCh)
+	go r.runWorker()
+}
+
+func (r *remoteCluster) runWorker() {
+	for {
+		keyObj, shutdown := r.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for cluster \"%q\" stopped", r.clusterID)
+			return
+		}
+		key := keyObj.(string)
+		func() {
+			defer r.queue.Done(key)
+			obj, exists, err := r.serviceInformer.GetIndexer().GetByKey(key)
+
+			if err != nil {
+				klog.Errorf("error retrieving service with key %q from the cache: %v", key, err)
+				// requeue the item to work on later
+				r.queue.AddRateLimited(key)
+				return
+			}
+			if !exists {
+				err := r.serviceDeleted(key)
+				if err != nil {
+					klog.Errorf("error deleting service %q in cluster %q: %v", key, r.clusterID, err)
+					r.queue.AddRateLimited(key)
+					return
+				}
+			} else {
+				err := r.serviceCreated(obj, key)
+				if err != nil {
+					klog.Errorf("error creating service %q in cluster %q: %v", key, r.clusterID, err)
+					r.queue.AddRateLimited(key)
+					return
+				}
+			}
+			r.queue.Forget(key)
+		}()
+	}
+}
+
+func (r *remoteCluster) serviceCreated(obj interface{}, key string) error {
+	klog.V(2).Infof("In remoteCluster serviceCreated for cluster %q, service %#v", r.clusterID, obj)
+	service := obj.(*core_v1.Service)
+	r.remoteServicesMutex.Lock()
+	defer r.remoteServicesMutex.Unlock()
+	if rs, exists := r.remoteServices[key]; !exists {
+		r.remoteServices[key] = r.createLightHouseCRD(service, r.clusterID)
+		klog.Infof("Creating a new MultiClusterService with service %q from cluster %q", key, r.clusterID)
+	} else {
+		klog.Infof("Adding service from cluster %q to existing MultiClusterService %q", r.clusterID, key)
+		rs.Spec.Items = append(rs.Spec.Items, r.createClusterServiceInfo(service, r.clusterID))
```

Comment:
Now that you've implemented `OnUpdate`, an element for `r.clusterID` may already exist so you don't want to blindly append. First search for the element, ie:

```
for _, info := range rs.Spec.Items {
    if info.ClusterID == r.clusterID {
        return nil
    }
}

rs.Spec.Items = append(...)
```

Reply:
See comment by tpantelis ^